### PR TITLE
Add Test cases for EndpointUtilty, LifetimeManager and ServiceConnection

### PR DESCRIPTION
**Pull request by zackliu**

### CODE REVIEW

File: src/Microsoft.Azure.SignalR/HubHost/IClientConnectionManager.cs [hunk 1]
```diff
@@ -2,7 +2,9 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Collections.Concurrent;
+using System.Runtime.CompilerServices;
 
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")]
```
[the remark this answers is not in the document]
Comment (ccic) [reply src/Microsoft.Azure.SignalR/HubHost/IClientConnectionManager.cs]: You do not need to create AssemblyInfo.cs. Another way is to include the internal interface into your test project, like this.

  <ItemGroup>
    <Compile Include="..\..\src\Microsoft.Azure.SignalR\HubHost\IClientConnectionManager.cs" Link="IClientConnectionManager.cs" />
  </ItemGroup>

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/ServiceConnectionProxy.cs [hunk 2]
```diff
@@ -170,9 +184,55 @@ public void RemoveClientConnection(string connectionId)
             }
         }
 
+        private async Task ProcessIncomingCoreAsync(PipeReader input, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                var result = await input.ReadAsync(cancellationToken);
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/ServiceConnectionProxy.cs]: break when `result.IsCanceled`?

File: test/Microsoft.Azure.SignalR.Tests/ServiceLifetimeManagerIT.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+            _serviceLifetimeManager = new ServiceLifetimeManager<TestHub>(serviceConnectionManager,
+                _proxy.ClientConnectionManager,
+                new DefaultHubProtocolResolver(new IHubProtocol[] { new JsonHubProtocol(), new MessagePackHubProtocol() }, NullLogger<DefaultHubProtocolResolver>.Instance),
+                NullLogger<ServiceLifetimeManager<TestHub>>.Instance
+                );
+        }
+
+        [Fact]
+        public async void SendAllAsync()
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/ServiceLifetimeManagerIT.cs]: We can write a `theory` and pass parameters for different scenarios, right?

File: test/Microsoft.Azure.SignalR.Tests/ServiceLifetimeManagerFacts.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            _serviceConnectionManagerMock = new Mock<IServiceConnectionManager<TestHub>>();
+            var clientConnectionManagerMock = new Mock<IClientConnectionManager>();
+            _serviceLifetimeManager = new ServiceLifetimeManager<TestHub>(_serviceConnectionManagerMock.Object,
+                clientConnectionManagerMock.Object,
+                new DefaultHubProtocolResolver(new IHubProtocol[] { new JsonHubProtocol(), new MessagePackHubProtocol() }, NullLogger<DefaultHubProtocolResolver>.Instance),
+                NullLogger<ServiceLifetimeManager<TestHub>>.Instance);
+        }
+
+        [Fact]
+        public async void SendAllAsync()
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/ServiceLifetimeManagerFacts.cs]: `theory` works here too.

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/ServiceConnectionProxy.cs [hunk 5]
```diff
@@ -27,12 +27,13 @@ internal class ServiceConnectionProxy : IClientConnectionManager, IClientConnect
...
         public TestConnection ConnectionContext { get; }
 
-        private ServiceConnection ServiceConnection { get; }
+        public ServiceConnection ServiceConnection { get; }
 
         public ConcurrentDictionary<string, ServiceConnectionContext> ClientConnections => ClientConnectionManager.ClientConnections;
 
         private readonly ConcurrentDictionary<string, TaskCompletionSource<ConnectionContext>> _waitForConnectionOpen = new ConcurrentDictionary<string, TaskCompletionSource<ConnectionContext>>();
         private readonly ConcurrentDictionary<string, TaskCompletionSource<object>> _waitForConnectionClose = new ConcurrentDictionary<string, TaskCompletionSource<object>>();
+        private readonly ConcurrentDictionary<string, TaskCompletionSource<object>> _waitForSpecificMessage = new ConcurrentDictionary<string, TaskCompletionSource<object>>();
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/ServiceConnectionProxy.cs]: Using `Type` as key is enough?

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/ServiceConnectionProxy.cs [hunk 6]
```diff
@@ -27,12 +27,13 @@ internal class ServiceConnectionProxy : IClientConnectionManager, IClientConnect
...
         public TestConnection ConnectionContext { get; }
 
-        private ServiceConnection ServiceConnection { get; }
+        public ServiceConnection ServiceConnection { get; }
 
         public ConcurrentDictionary<string, ServiceConnectionContext> ClientConnections => ClientConnectionManager.ClientConnections;
 
         private readonly ConcurrentDictionary<string, TaskCompletionSource<ConnectionContext>> _waitForConnectionOpen = new ConcurrentDictionary<string, TaskCompletionSource<ConnectionContext>>();
         private readonly ConcurrentDictionary<string, TaskCompletionSource<object>> _waitForConnectionClose = new ConcurrentDictionary<string, TaskCompletionSource<object>>();
+        private readonly ConcurrentDictionary<Type, TaskCompletionSource<object>> _waitForSpecificMessage = new ConcurrentDictionary<Type, TaskCompletionSource<object>>();
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/ServiceConnectionProxy.cs]: `_waitForSpecificMessage` -> `_waitForMessage`

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestClientConnectionManager.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Azure.SignalR.Tests.Infrastructure
+{
+    class TestClientConnectionManager : IClientConnectionManager
+    {
+        public void AddClientConnection(ServiceConnectionContext clientConnection)
+        {
+            return;
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestClientConnectionManager.cs]: Remove unnecessary code.

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestClientConnectionManager.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        }
+
+        public void RemoveClientConnection(string connectionId)
+        {
+            return;
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestClientConnectionManager.cs]: remove

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestServiceConnectionManager.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Azure.SignalR.Protocol;
+
+namespace Microsoft.Azure.SignalR.Tests.Infrastructure
+{
+    class TestServiceConnectionManager<THub> : IServiceConnectionManager<THub> where THub : Hub
+    {
+        private readonly ConcurrentDictionary<Type, int> _writeAsyncCallCount = new ConcurrentDictionary<Type, int>();
+        public void AddServiceConnection(ServiceConnection serviceConnection)
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestServiceConnectionManager.cs]: Empty line between fields and method.

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestServiceConnectionManager.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            return;
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestServiceConnectionManager.cs]: Remove.

File: test/Microsoft.Azure.SignalR.Tests/ServiceEndpointUtilityFacts.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+using System.Security.Claims;
+using Microsoft.Azure.SignalR.Tests.Infrastructure;
+using Microsoft.Extensions.Options;
+using Xunit;
+
+namespace Microsoft.Azure.SignalR.Tests
+{
+    public class ServiceEndpointUtilityFacts
+    {
+        private static readonly string Endpoint = "https://myendpoint";
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/ServiceEndpointUtilityFacts.cs]: -> `private const string`

File: test/Microsoft.Azure.SignalR.Tests/ServiceEndpointUtilityFacts.cs [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+        private static readonly string AccessKey = "nOu3jXsHnsO5urMumc87M9skQbUWuQ+PE5IvSUEic8w=";
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/ServiceEndpointUtilityFacts.cs]: same here

File: test/Microsoft.Azure.SignalR.Tests/ServiceEndpointUtilityFacts.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+            string endpoint = EndpointUtility.GetClientEndpoint<TestHub>();
+            Assert.Equal(endpoint, $"{Endpoint}:5001/client/?hub={nameof(TestHub).ToLower()}");
+        }
+
+        [Fact]
+        public void GenerateServerAccessToken()
+        {
+            string userId = "UserA";
+            var token = EndpointUtility.GenerateServerAccessToken<TestHub>(userId);
+            var expireDate = new JwtSecurityTokenHandler().ReadJwtToken(token).ValidTo;
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/ServiceEndpointUtilityFacts.cs]: use a static `JwtSecurityTokenHandler`

File: test/Microsoft.Azure.SignalR.Tests/ServiceLifetimeManagerFacts.cs [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        {
+            var serviceConnectionManager = new TestServiceConnectionManager<TestHub>();
+            var serviceLifetimeManager = new ServiceLifetimeManager<TestHub>(serviceConnectionManager,
+                new TestClientConnectionManager(),
+                new DefaultHubProtocolResolver(new IHubProtocol[] { new JsonHubProtocol(), new MessagePackHubProtocol() }, NullLogger<DefaultHubProtocolResolver>.Instance),
+                NullLogger<ServiceLifetimeManager<TestHub>>.Instance);
+
+            await CallFunction(serviceLifetimeManager, functionName);
+
+            Assert.Equal(1, serviceConnectionManager.GetCallCount(type));
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/ServiceLifetimeManagerFacts.cs]: Should also verify message fields, right?

**Comment by xscript** [comment]
@zackliu Travis build of your latest commit failed. Need to fix it.

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/IHandshackMessageFactory.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.SignalR.Protocol;
+
+namespace Microsoft.Azure.SignalR.Tests.Infrastructure
+{
+    interface IHandshackMessageFactory
+    {
+        ServiceMessage GetHandshackResposeMessage();
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/IHandshackMessageFactory.cs]: `Handshack` -> `Handshake`. Update all of them in changed files.

Comment (xscript) [reply test/Microsoft.Azure.SignalR.Tests/Infrastructure/IHandshackMessageFactory.cs]: `Respose` -> `Response`

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/PipeReaderExtensions.cs [hunk 16]
```diff
@@ -1,3 +1,6 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/PipeReaderExtensions.cs]: 👍

File: test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestClientConnectionManager.cs [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Azure.SignalR.Tests.Infrastructure
+{
+    class TestClientConnectionManager : IClientConnectionManager
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestClientConnectionManager.cs]: We can just use `ClientConnectionManager` right? So this class is not necessary.

File: test/Microsoft.Azure.SignalR.Tests/ServiceConnectionFacts.cs [hunk 18]
```diff
@@ -162,12 +176,176 @@ public class ServiceConnectionFacts
...
+
         private async Task<T> ReadServiceMessageAsync<T>(PipeReader input, int timeout = 5000) where T : ServiceMessage
         {
             var data = await input.ReadSingleAsync().OrTimeout(timeout);
             var buffer = new ReadOnlySequence<byte>(data);
             Assert.True(Protocol.TryParseMessage(ref buffer, out var message));
             return Assert.IsType<T>(message);
         }
+
+        private class ConnectionFactoryForReconnection : IConnectionFactory
```
Comment (xscript) [inline test/Microsoft.Azure.SignalR.Tests/ServiceConnectionFacts.cs]: Let's discuss offline to find a cleaner implementation.

File: src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManager.cs [hunk 19]
```diff
@@ -15,6 +15,10 @@ namespace Microsoft.Azure.SignalR
 {
     internal class ServiceLifetimeManager<THub> : HubLifetimeManager<THub> where THub : Hub
     {
+        private const string MarkerNotConfiguredError =
+            "'UseAzureSignalR(...)' not called after calling 'AddAzureSignalR(...)'." +
```
Comment (davidfowl) [inline src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManager.cs]: > UseAzureSignalR(...)' not called after calling 'AddAzureSignalR(...). Please always use 'AddAzureSignalR(...)' and 'UseAzureSignalR(...)' altogether.

Change this to.

AddAzureSignalR(...) was called without matching call to IApplicationBuilder.UseAzureSignalR(...).

We should also point to an FWLink that points to the docs on how to configure the service.

Comment (xscript) [reply src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManager.cs]: Sure.
We don't have docs specifically for configuring the service. There is supposed to be a FAQ page in docs.microsoft.com. We can put this issue in there. But it is not ready yet. I will ask what is the progress.

**Pull request merged**